[DNM][Downstream][hotfix]mon/Elector: Change how we handle removed_ranks and notify_rank_removed()

kamoltat opened the pull request:
Problem:

Currently, there is an issue when performing a DR test with 2 sites stretch cluster
where removing monitors and adding new ones to the cluster
causes incorrect ``rank`` in ConnectionTracker class.

This causes the monitor to think that they are someone else
in the ``peer_tracker`` copy and will never update the
correct field of itself, causing a deadlock in the election process (Ceph becoming unresponsive)
when using election strategy: 3 (Connectivity mode).

Solution:

1. It was really hard to debug the issue so the first thing we did was to add additional loggings to ConnectionTracker,
Elector and ElectionLogic Classes.

2. In `ConnectionTracker::receive_peer_report` we loop through ranks which is bad when there is `notify_rank_removed` before this and the ranks are not adjusted yet. When we rely on the rank in certain scenarios, we end up with extra peer_report copy which we don't want. Therefore, instead of passing `report.rank` in the function
`ConnectionTracker::reports`, we pass `i.first` instead so that we trim old ranks properly. We also added an assert in notify_rank_removed(), comparing the expected rank provided by the monmap against the rank that we adjust ourselves to as a sanity check. We edited test/mon/test_election.cc to reflect the changes made in notify_rank_removed().

3. MonMap::removed_rank does not get cleared every update of the epoch, this was the root cause of the problem. Therefore, we fix this by making MonMap clear removed_ranks every update. Moreover, When there is discontinuity between monmaps for more than 1 epoch or the new monitor never joined the quorum before,
we always reset `peer_tracker`.

4. Added a way for us to manually reset `peer_tracker.rank` when executing the command: `ceph connection scores reset` for each monitor. The peer_tracker.rank will match the current rank of the Monitor.

5. When upgrading the monitors (including booting up), we check if `peer_tracker` is dirty or not. If so, we clear it. Added some functions in the `Elector` and `ConnectionTracker` classes to check for clean `peer_tracker`. Moreover, there could be some cases where due to startup weirdness or abnormal circumstances, we might get a report from our own rank. Therefore, it doesn't hurt to add a sanity check in `ConnectionTracker::report_live_connection` and `ConnectionTracker::report_dead_connection`.

Fixes: https://tracker.ceph.com/issues/58049

Signed-off-by: Kamoltat <ksirivad@redhat.com>
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

kamoltat commented:
Hi @gregsfortytwo 

I have tested this by running my own reproducer local using vstart.sh, removing and adding monitors in stretch_cluster.
So far I was able to come up with this fix to prevent peer_tracker.rank from being adjusted incorrectly and for ``removed_ranks`` to store old state so that we don't incorrectly remove ranks and adjust them.

Please only look at the commits that have actual code changes and not just logging:

9c2af4d2aff803f443f0bceaeb2acf05d54e4d22
ebe20d9c8dc91962c8d10f8247fd42476cc4c684
b223a71be20d9caedd602c0d9beb41d605b56d49

Also, I was thinking of separating elector.notify_rank_removed() from notify_new_monmap().
So that we don't have to pass extra flags in notify_new_monmap() when we don't want to call notify_rank_removed().
However, I don't want to overcomplicate things and I have more peace of mind when I don't change too many things.

gregsfortytwo commented on src/mon/MonmapMonitor.cc:
```diff
@@ -129,7 +129,9 @@ void MonmapMonitor::create_pending()
   pending_map.epoch++;
   pending_map.last_changed = ceph_clock_now();
-  dout(10) << __func__ << " monmap epoch " << pending_map.epoch << dendl;
+  pending_map.removed_ranks.clear();
```
...yep, this was definitely supposed to happen here and not doing so was a big screw-up on my part. :/ Nice find!

If you look at the git history you'll see I said I expected this to always be of size 1, but I made it a container instead of an integer just in case I was missing some way for multiple removals to happen in a single map...really got that wrong. I should have made it an integer and asserted it wasn't set whenever a mon removal happened instead; that would have caught this and made it safe early. :/

kamoltat replied:
I see, I actually prefer we change the data type to int, However, because we are gearing towards applying the patch as an update for the customer's existing cluster I'm not sure if changing the data type will mess things up or not.

gregsfortytwo commented on src/mon/MonmapMonitor.cc:
```diff
@@ -129,7 +129,9 @@ void MonmapMonitor::create_pending()
-  dout(10) << __func__ << " monmap epoch " << pending_map.epoch << dendl;
+  pending_map.removed_ranks.clear();
+  dout(10) << __func__ << " monmap epoch " << pending_map.epoch 
+    << " removed_ranks " << pending_map.removed_ranks << dendl;
```
...but this log line probably needs to go before clearing the removed_ranks? I'm not quite sure what the point of this change is, unless you were debugging to discover the removed_ranks were still present?

gregsfortytwo commented on src/mon/MonmapMonitor.cc:
```diff
@@ -761,6 +763,8 @@ bool MonmapMonitor::prepare_command(MonOpRequestRef op)
     pending_map.disallowed_leaders.erase(name);
     pending_map.last_changed = ceph_clock_now();
+    dout(20) << "removing mon." << name << " at " << addrs
+       << ", there will be " << pending_map.size() << " monitors" << dendl;
```
If you want this output in the log, you should dump a copy from the stringstream that renders it in the next line instead of re-generating it. :) (ie, ```dout(20) << ss.str() << dendl;``` I think should do it.)

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -2006,7 +2006,12 @@ void Monitor::handle_probe_reply(MonOpRequestRef op)
-      notify_new_monmap(false);
+      if (!has_ever_joined) {
+        dout(20) << " !has_ever_joined " << *monmap << dendl;
+        notify_new_monmap(false, false);
```
Hmm. I forget when we start pinging and generating ConnectionTracker state. Is the following possible?

New monitor turns on
Start pinging existing 5 monitors
Start copying paxos state from mon.1
mon.4 is removed from monitor list
finish copying state
get new map with mon.4 in removed_ranks
fail to remove because we are !has_ever_joined

kamoltat replied:
Hi Greg,

How about we only allow mon remove when it has joined quorum before, for other cases like
never joined or joined but we have a very outdated version, we don't adjust and also make
sure to nuke peer_report, incase it has already been populated with like a removed_rank:

````
      if (newmap->get_epoch() - monmap->get_epoch() == 1 &&
              has_ever_joined) {
        notify_new_monmap(false);
      } else {
        notify_new_monmap(false, false);
        elector.notify_clear_peer_state();
      }
````

gregsfortytwo replied:
I don't think we can block mon removal on stuff like that — what if there's physical hardware loss between adding it to the map and when it actually comes on?
And that wouldn't fix the scenario I proposed either — the problem is just having a new monitor see a different removed monitor at a bad time.

kamoltat replied:
by state, do you mean `peer_tracker` in the elector? if so I guess you mean even if we are 1 epoch apart we should always `elector.notify_clear_peer_state();`

gregsfortytwo replied:
Right, if we see a discontinuity we need to reset our scores because we don't know what might have happened to the monitor set and how the scores should align.

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -2006,7 +2006,12 @@ void Monitor::handle_probe_reply(MonOpRequestRef op)
+        notify_new_monmap(false, false);
+      } else {
+        notify_new_monmap(false);
+      }
```
I'm a little concerned about hitting a similar issue here compared to the above branch.

If we've been dead and missed 3 monmap epochs, the removed_ranks member is just junk to us. I think if we have a map discontinuity, we need to reset our local tracking.

kamoltat replied:
Resolved by going into elector.notify_clear_peer_state(); no matter what

gregsfortytwo reviewed (changes requested):
Good find! There's at least one edge case still not covered, and a few others I'm worried about.

Plus of course figuring out what we want to do with all the new logging — much of it probably has value but at 20 or 30 instead of 10?

kamoltat commented:
Hi @gregsfortytwo

thank you for the comments, I will get to them shortly.

I just pushed a commit that rewrites ConnectionTracker::notify_rank_removed():

032629c668cc549448fa39cb10c724f7c4df0ce6

the motivations are in the commit message.

But in short, I believe by doing it this way will make the peer_tracker cleaner and less prone to
getting into a deadlock state during the election.

I have tested this in my local vstart, removing the lowest rank, highest rank, middle and etc.

kamoltat commented on src/mon/ConnectionTracker.cc:
```diff
@@ -49,13 +57,17 @@ const ConnectionReport *ConnectionTracker::reports(int p) const
 
 void ConnectionTracker::receive_peer_report(const ConnectionTracker& o)
 {
+  dout(10) << __func__ << dendl;
```
Yes so most of the logs are for debugging purposes, I'll clean them up and set a lot of them for 30.

kamoltat commented on src/mon/ConnectionTracker.cc:
```diff
@@ -229,8 +230,10 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
+  dout(20) << "my_reports after: " << my_reports << dendl;
+  dout(20) << "peer_reports after: " << peer_reports << dendl;
+  dout(20) << "my rank after: " << rank << dendl;
+  ceph_assert(rank == new_rank);
```
This assert did not trigger when I tested locally with this patch

kamoltat commented:
@gregsfortytwo Just updated this PR with regards to your comments, I also added the mechanism where we can reset `peer_tracker.rank` to match `mon. rank`, this is embedded in the function: `notify_clear_peer_state()`. Which get triggered when we do `ceph connection scores reset`.

kamoltat commented:
jenkins test api

kamoltat commented:
jenkins test api

gregsfortytwo commented on src/mon/ConnectionTracker.cc:
```diff
@@ -203,46 +203,51 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
-    peer_reports.erase(pi++);
+  // Go through everyone, remove & adjust current and history map.
+  for (auto pi = peer_reports.begin(); pi != peer_reports.end(); pi++) {
+    peer_reports[pi->first].current.erase(rank_removed);
```
```pi->second.current.erase(rank_removed)```

kamoltat replied:
yep good catch!

gregsfortytwo commented on src/mon/ConnectionTracker.cc:
```diff
@@ -203,46 +203,51 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
+  // Go through everyone, remove & adjust current and history map.
+  for (auto pi = peer_reports.begin(); pi != peer_reports.end(); pi++) {
+    peer_reports[pi->first].current.erase(rank_removed);
+    peer_reports[pi->first].history.erase(rank_removed);
```
```pi->second.history.erase(rank_removed)```

etc etc in this loop. Is there a reason you're looking up the peer_report each time?

kamoltat replied:
nope, should just use pi->second

gregsfortytwo commented on src/mon/ConnectionTracker.cc:
```diff
@@ -203,46 +203,51 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
+        peer_reports[pi->first].history[hi->first - 1] = hi->second;
+        peer_reports[pi->first].current.erase(ci++);
+        peer_reports[pi->first].history.erase(hi++);
+    }
```
So I really, really don't like this. The peer_reports mechanism is designed so only the monitor generating a report updates the numbers, and then we assimilate them. Directly manipulating them like this is icky, and actually can lead to differing election outcomes during an upgrade from the old-way monitors to this way.

From the commit message, I think this is the real problem:
> However, there is a case where if we remove the highest rank of the quorum, then we will leave a the higest rank in report.current and report.history

If we just resolve that, which sounds like a fairly simple bug in ConnectionTracker::receive_peer_reports, I don't think we need to do this twiddling.

kamoltat replied:
I think even with the old code, we are already directly manipulating peer_reports by adjusting ranks: https://github.com/ceph/ceph/blob/main/src/mon/ConnectionTracker.cc#L171-L176.
my thoughts are that since we are already adjusting the ranks of peer_reports, might as well adjust the ranks of peer_reports.current and peer_reports.history as well so that they are aligned with each other.

I understand your concerns about manipulating them, but if we go with that, then we should just only remove and not adjust the ranks in peer_reports at all correct?

gregsfortytwo replied:
> I think even with the old code, we are already directly manipulating peer_reports by adjusting ranks: https://github.com/ceph/ceph/blob/main/src/mon/ConnectionTracker.cc#L171-L176. my thoughts are that since we are already adjusting the ranks of peer_reports, might as well adjust the ranks of peer_reports.current and peer_reports.history as well so that they are aligned with each other.
> 
> I understand your concerns about manipulating them, but if we go with that, then we should just only remove and not adjust the ranks in peer_reports at all correct?

That link is just removing scores, not updating the individual reports?

kamoltat replied:
The old code moves the peer_report down so in a way it is adjusting the ranks implicitly since functions like ``get_total_connection_score`` rely on the position of the peer_report rather than the actual peer_report.rank.

If we use the old code we'll have this in this scenario:

```
we are in mon.0 point of view.

notify_rank_removed: 2

peer_reports before remove 2: {
0=rank=0, current links: {1=1, 2=1, 3=1, 4=1}, history: {1=0.9, 2=0.9 3=0.9, 4=0.9},
1=rank=1, current links: {0=1, 2=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9},
2=rank=2, current links: {0=1, 1=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9}, 
3=rank=3, current links: {0=1, 1=1, 2=1, 4=1}, history: {0=0.8, 1=0.7, 2=0.6, 4=0.9},
4=rank=4, current links: {0=1, 1=1, 2=1, 3=1}, history: {0=0.9, 1=0.6, 2=0.5, 3=0.7},
}

peer_reports after remove 2: {
0=rank=0, current links: {1=1, 2=1, 3=1, 4=1}, history: {1=0.9, 2=0.9 3=0.9, 4=0.9},
1=rank=1, current links: {0=1, 2=1, 3=1, 4=1}, history: {0=0.9, 2=0.9, 3=0.9, 4=0.9}, 
2=rank=3, current links: {0=1, 1=1, 2=1, 4=1}, history: {0=0.8, 1=0.7, 2=0.6, 4=0.9},
3=rank=4, current links: {0=1, 1=1, 2=1, 3=1}, history: {0=0.9, 1=0.6, 2=0.5, 3=0.7},
}

```

So now, rank 4 becomes rank 3, and rank 3 becomes rank 2.

let's say mon.0 call get_total_connection_score peer 3,  peer_report[2] would give us nothing, but it should have given us 0.9.

So what I am trying to say is that since we are moving the ranks down by 1 already with the old code. I don't think it is fair to not move the ranks in current and history.

kamoltat replied:
Moreover, the fact that the old code doesn't adjust the actual peer_report.rank leaves us a vulnerability as well since let's say after we remove ranks with the old code and we ping another monitor and they happened to to `assimilate_connection_reports` by us, then with our unadjusted peer_report.rank, we will populate the incorrect information on their peer_reports and this also can lead to differing election outcomes.

kamoltat replied:
@gregsfortytwo PTAL

gregsfortytwo commented on src/mon/MonMap.cc:
```diff
@@ -357,7 +357,7 @@ void MonMap::print_summary(ostream& out) const
     has_printed = true;
   }
-  out << "}";
+  out << "}" << " removed_ranks: {" << removed_ranks << "}";
```
I'd expect the addition of this log line to break some unit tests, which probably need to get their output-text-matching files updated to match the new format.

kamoltat replied:
I see, I'll try to find a way to output the `removed_ranks` better

gregsfortytwo replied:
This is the right way. I think you just need to update some of the out.t test files.

Or maybe we don't actually check the Mon map output in any of those? Which I would be surprised about...

...maybe those tests all use the dump() output instead? Actually we should probably put this there instead of just in print_summary? Would be a bit weird for the summary to have info not in the dump. :)

kamoltat replied:
make sense

gregsfortytwo commented on src/mon/ConnectionTracker.cc:
```diff
@@ -106,6 +106,10 @@ void ConnectionTracker::report_live_connection(int peer_rank, double units_alive
+  if (peer_rank == rank) {
+    lderr(cct) << "Got a report from my own rank, hopefully this is startup weirdness, dropping" << dendl;
+    return;
+  }
```
Have you seen this error occur? I don't understand how it would happen and that scares me.

gregsfortytwo replied:
Ah, that makes sense.

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -950,6 +950,9 @@ int Monitor::init()
 
   state = STATE_PROBING;
+
+  elector.notify_clear_peer_state();
```
And I don't really get why we want to unconditionally clear peer state here. Doesn't...doesn't this throw away our connection scores even time the monitor boots? That's not what we want!

kamoltat replied:
My motivation for putting that there is for the case where we want to upgrade from a buggy version of Ceph where the Connection Scores are messed up to this current version, we don't have to rely on the user/admin to execute the command:

``ceph daemon connection scores reset``

I thought this was the best place to put it because the function is in the code path for when we upgrade the monitor. I know it overlaps with the scenario where the monitor is down or restart, but at the time I couldn't find a better place to put it. I don't think it will hurt to clear the scores and also reassign the rank of the peer_tracker to the actual rank according to Monmap every time the Monitor restarts.

In my opinion, I think the benefits here outweigh the cons when we throw away the scores and populate a new one since I think it is important for the patch to make sure we don't retain any issues from the previous buggy version when upgrading to this patch.

But ofc you have a better solution, I'm all ears.

gregsfortytwo replied:
I'm not sure the right place to fix this, but as we said verbally, I don't think we can toss out the scores on every monitor boot. Besides being messy, I'm worried about bad behavior in degenerate cases.

kamoltat replied:
Yep will add a condition to only clear the peer_tracker if it is messed up beyond recovery.

kamoltat commented:
@gregsfortytwo Just pushed the new changes addressing your concerns except for the `Directly manipulating peer_reports`, I've outlined my explanation of why I think we should adjust the current and history, let me know what you think about my reasons. Thank you

anthonyeleven reviewed (commented):
I don't see doc content here, so abstaining

idryomov reviewed (commented):
Why is this targeted for pacific?  Is there a corresponding PR for main?

kamoltat commented:
@idryomov this is for a downstream hotfix, we open it for reference and code review it's a DNM for upstream

kamoltat commented on src/mon/ConnectionTracker.cc:
```diff
@@ -219,22 +216,27 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
-    peer_reports[pi->first - 1] = pi->second;
-    peer_reports.erase(pi++);
+    peer_reports[pi->first - 1] = pi->second; // copy content of next rank to ourself.
+    --peer_reports[pi->first - 1].rank; // update the rank of our new content.
```
If we don't also update the rank of our peer locally after we move our peers, then it could be a problem so that's why I'm also updating the rank

gregsfortytwo replied:
Can you explain this "could be a problem"?
I'm not sure futzing with these ranks is an issue, but I really want them to be read-only. I did some reasonably solid proofs on convergence with gossip-based updates to these but it was all based on treating the peer_reports as a vector clock thing.

kamoltat replied:
if we don't want to change things in peer_reports, then we might have to change how we do things in
ConnectionTracker::reports or ConnectionTracker::receive_peer_report. What we might want to do is instead of passing `report.rank`, we can pass `i.first` (the key of peer_reports map)

kamoltat replied:
Sounds good, Addressed in new push

kamoltat commented:
jenkins test api

gregsfortytwo commented on src/mon/ConnectionTracker.cc:
```diff
@@ -243,6 +245,8 @@ void ConnectionTracker::notify_rank_removed(int rank_removed)
   //check if the new_rank from monmap is equal to our adjusted rank.
   ceph_assert(rank == new_rank);
+
+  increase_version();
```
oh yes, these are good.

gregsfortytwo commented on src/mon/ConnectionTracker.h:
```diff
@@ -185,6 +185,18 @@ class ConnectionTracker {
+  bool is_clean(int mon_rank) {
+    if (rank != mon_rank ||
+      my_reports.rank != mon_rank) {
+      return 0;
```
Nit: the function is declared as a bool; we should return a bool instead of relying on ints being cast.

kamoltat replied:
Addressed in new push

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -950,6 +950,12 @@ int Monitor::init()
   state = STATE_PROBING;
+
+  if (!elector.peer_tracker_is_clean()){
+    dout(10) << "peer_tracker is not clean, clearing ..." << dendl;
```
Let's call this something besides "clean". Could say "consistent", or "peer_tracker looks inconsistent due to previous bad update logic; clearing..."

kamoltat replied:
Addressed in new push

gregsfortytwo commented on src/mon/Elector.h:
```diff
@@ -357,6 +357,10 @@ class Elector : public ElectionOwner, RankProvider {
    */
   void start_participating();
+  /**
+  * Check if our peer_tracker is clean
```
s/is clean/is self-consistent, not suffering from [tracker URL]/

kamoltat replied:
Addressed in new push

gregsfortytwo commented on src/mon/ConnectionTracker.h:
```diff
@@ -185,6 +185,18 @@ class ConnectionTracker {
+  */
+  bool is_clean(int mon_rank) {
+    if (rank != mon_rank ||
+      my_reports.rank != mon_rank) {
```
Is this the condition we're always failing? I thought it could also just involve having some extra peer_reports in the map.

gregsfortytwo replied:
Yeah; I'm worried about the case where some monitors have an old rank 6 score sitting around and others don't.

kamoltat replied:
Addressed in new push

gregsfortytwo commented on src/mon/Monitor.cc:
```diff
@@ -2004,9 +2004,17 @@ void Monitor::handle_probe_reply(MonOpRequestRef op)
+        notify_new_monmap(false, false);
+      }
+
+      elector.notify_clear_peer_state();
```
If we're unconditionally clearing the state, there's not much point to the dance trying to identify the discontinuity above. Shouldn't we just clear it in the "else" branch above and let it be otherwise, though?

kamoltat replied:
Addressed in new push

gregsfortytwo commented:
Also I notice this is targeted at pacific and titled for downstream, but these fixes certainly need to go into main and quincy as well...

gregsfortytwo commented:
LGTM!

kamoltat commented:
jenkins test make check

kamoltat commented:
jenkins retest this please

kamoltat commented:
make check failure unrelated, rbd-unit-tests are known to be flaky 

 12 - run-rbd-unit-tests-1.sh (Failed)

kamoltat commented:
jenkins test make check

kamoltat commented:
jenkins test api

ljflores commented:
@kamoltat let me know when this is ready for testing. You can tag it with `needs-qa` and ping me.

kamoltat commented:
> @kamoltat let me know when this is ready for testing. You can tag it with `needs-qa` and ping me.

@ljflores thanks Laura, it's ready.

amathuria commented:
@kamoltat  one of the failures in the run had the following assert failure: 
/a/yuriw-2022-11-28_21:10:48-rados-wip-yuri10-testing-2022-11-28-1042-pacific-distro-default-smithi/7095280

Raised a tracker for it here: https://tracker.ceph.com/issues/58114

  ```
  -1> 2022-11-29T04:58:26.511+0000 7fc090d88700 -1 /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.2.10-1303-g8622b733/rpm/el8/BUILD/ceph-16.2.10-1303-g8622b733/src/mon/ConnectionTracker.cc: In function 'void ConnectionTracker::notify_rank_removed(int, int)' thread 7fc090d88700 time 2022-11-29T04:58:26.510305+0000
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.2.10-1303-g8622b733/rpm/el8/BUILD/ceph-16.2.10-1303-g8622b733/src/mon/ConnectionTracker.cc: 252: FAILED ceph_assert(rank == new_rank)

 ceph version 16.2.10-1303-g8622b733 (8622b7333736b9290df89221ec0bb0873f4194dd) pacific (stable)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x158) [0x7fc09e4b3df4]
 2: /usr/lib64/ceph/libceph-common.so.2(+0x27900e) [0x7fc09e4b400e]
 3: (ConnectionTracker::notify_rank_removed(int, int)+0xd68) [0x560d07e4d6c8]
 4: (Elector::notify_rank_removed(int, int)+0x123) [0x560d07e3b053]
 5: (Monitor::notify_new_monmap(bool, bool)+0x2a1) [0x560d07da3ee1]
 6: (MonmapMonitor::update_from_paxos(bool*)+0x994) [0x560d07df1b24]
 7: (PaxosService::refresh(bool*)+0x10e) [0x560d07ed01fe]
 8: (Monitor::refresh_from_paxos(bool*)+0x18c) [0x560d07d7a3ac]
 9: (Paxos::do_refresh()+0x57) [0x560d07ec26a7]
 10: (Paxos::handle_commit(boost::intrusive_ptr<MonOpRequest>)+0x309) [0x560d07ec2b39]
 11: (Paxos::dispatch(boost::intrusive_ptr<MonOpRequest>)+0x457) [0x560d07ecafe7]
 12: (Monitor::dispatch_op(boost::intrusive_ptr<MonOpRequest>)+0x1324) [0x560d07db62d4]
 13: (Monitor::_ms_dispatch(Message*)+0x670) [0x560d07db6d60]
 14: (Dispatcher::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x5c) [0x560d07de5f0c]
 15: (DispatchQueue::entry()+0x126a) [0x7fc09e6fa5ea]
 16: (DispatchQueue::DispatchThread::entry()+0x11) [0x7fc09e7ad921]
 17: /lib64/libpthread.so.0(+0x814a) [0x7fc09c1e614a]
 18: clone()

```

It looked related to this change. The coredump is available as well: 
/a/yuriw-2022-11-28_21:10:48-rados-wip-yuri10-testing-2022-11-28-1042-pacific-distro-default-smithi/7095280/remote/smithi203/coredump/

kamoltat commented:
@amathuria thanks for letting me know, looking at it right now

kamoltat commented:
I have investigated this thoroughly and I have concluded that this failure is due to how we are not properly using
CEPH_ARGS+="--mon-initial-members=a,b,c " in the standalone test which results in prepopulated "removed_ranks" which causes the assert. In summary, it's the standalone test's fault, I'm writing up the details as we speak will share it ASAP.

kamoltat commented:
Analysis on my investigation:
https://docs.google.com/document/d/1Rk-PUZ6cwBT8a9Z5RAFY5sb8-xFx-K02k7l4T2p2U5Q/edit

kamoltat commented:
@gregsfortytwo I have decided to get rid of the --mon-initial-members setting in the standalone test, as I have concluded that it does not really play a role in the tests at all except pre-populated ``removed_ranks`` with members that they think were originally there but it was a result of monmap initialization based on address/host provided by --mon-host setting. The details can be found in the commit: https://github.com/ceph/ceph/pull/48698/commits/b143cc93c733b3df24ec8f6eecb83872a41ee18d

kamoltat commented:
jenkins test api

kamoltat commented:
Closing and moving this PR to https://github.com/ceph/ceph/pull/49312
For backport cleanliness of upstream PRs.
